### PR TITLE
Exclude with statement names from tables

### PR DESCRIPTION
**Pull request by willmostly**
## Description
Exclude with statement names from the list of tables extracted from SQL by `TrinoQueryProperties`

## Additional context and related issues
The existing TrinoQueryProperties will include WITH statement names in the list of tables. For example, for the query 
```
WITH arbitrary_name AS (
SELECT c1 from c.s.t)
SELECT * FROM arbutrary_name
```
the tables list will be `{"c.s.t", "arbitrary_name"}`, or, if default catalog and schema are not specified, processing of the SQL will fail when `TrinoQueryPoperties` attempts to qualify `arbitrary_name` with a catalog and schema.

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
* Support `trinoQueryProperties` for SQL containing `WITH` statements
```

### CODE REVIEW

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java [hunk 1]
```diff
@@ -194,13 +196,25 @@ else if (statement instanceof ExecuteImmediate executeImmediate) {
 
             queryType = statement.getClass().getSimpleName();
             resourceGroupQueryType = StatementUtils.getQueryType(statement).toString();
-            ImmutableSet.Builder<QualifiedName> tableBuilder = ImmutableSet.builder();
+            List<QualifiedName> allQUalifiedNames = new ArrayList<>();
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:199]: Why did you change to ArrayList?  #Also, the new name has a typo. `U` should be a lowercase.

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java [hunk 2]
```diff
@@ -260,27 +274,30 @@ private String decodePreparedStatementFromHeader(String headerValue)
         return new String(preparedStatement, UTF_8);
     }
 
-    private void getNames(Node node, ImmutableSet.Builder<QualifiedName> tableBuilder,
+    private void getNames(
+            Node node,
+            List<QualifiedName> allQualifiedNames,
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:279]: I don't understand the difference between `tableBuilder` and `allQualifiedNames`. What's the motivation of this renaming?

File: gateway-ha/src/test/java/io/trino/gateway/ha/router/TestRoutingGroupSelector.java [hunk 3]
```diff
@@ -404,6 +404,36 @@ void testTrinoQueryPropertiesTableExtraction(String query, Set<String> catalogs,
...
     }
 
+    @Test
+    void testWithQueryNameExcluded()
+            throws IOException
+    {
+        String query = """
+                WITH dos (c1) as (select c1 from uno),
+                uno(c1) as (select c1 from cat.schem.tbl1)
+                SELECT c1 from uno, dos
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/router/TestRoutingGroupSelector.java:414]: Uppercase as, select, from.

File: gateway-ha/src/test/java/io/trino/gateway/ha/router/TestRoutingGroupSelector.java [hunk 4]
```diff
@@ -404,6 +404,36 @@ void testTrinoQueryPropertiesTableExtraction(String query, Set<String> catalogs,
...
+        BufferedReader bufferedReader = new BufferedReader(new StringReader(query));
+        HttpServletRequest mockRequestWithDefaults = prepareMockRequest();
+        when(mockRequestWithDefaults.getReader()).thenReturn(bufferedReader);
+        when(mockRequestWithDefaults.getHeader(TrinoQueryProperties.TRINO_CATALOG_HEADER_NAME)).thenReturn(DEFAULT_CATALOG);
+        when(mockRequestWithDefaults.getHeader(TrinoQueryProperties.TRINO_SCHEMA_HEADER_NAME)).thenReturn(DEFAULT_SCHEMA);
+
+        TrinoQueryProperties trinoQueryPropertiesWithDefaults = new TrinoQueryProperties(mockRequestWithDefaults, requestAnalyzerConfig);
+
+        assertThat(trinoQueryPropertiesWithDefaults.tablesContains(String.format("%s.%s.%s", DEFAULT_CATALOG, DEFAULT_SCHEMA, "uno"))).isFalse();
+        assertThat(trinoQueryPropertiesWithDefaults.tablesContains(String.format("%s.%s.%s", DEFAULT_CATALOG, DEFAULT_SCHEMA, "does"))).isFalse();
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/router/TestRoutingGroupSelector.java:425]: Static import `format` method. https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#prefer-string-formatting

File: gateway-ha/src/test/java/io/trino/gateway/ha/router/TestRoutingGroupSelector.java [hunk 5]
```diff
@@ -404,6 +404,36 @@ void testTrinoQueryPropertiesTableExtraction(String query, Set<String> catalogs,
...
+        assertThat(trinoQueryPropertiesWithDefaults.tablesContains("cat.schem.tbl1")).isTrue();
+
+        HttpServletRequest mockRequestNoDefaults = prepareMockRequest();
+        bufferedReader = new BufferedReader(new StringReader(query));
+        when(mockRequestNoDefaults.getReader()).thenReturn(bufferedReader);
+
+        TrinoQueryProperties trinoQueryPropertiesNoDefaults = new TrinoQueryProperties(mockRequestNoDefaults, requestAnalyzerConfig);
+        assertThat(trinoQueryPropertiesNoDefaults.getTables().size()).isEqualTo(1);
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/router/TestRoutingGroupSelector.java:433]: AssertJ provides `hasSize` method.

**Comment by mosabua** [comment]
This overlaps with #531 ...

File: gateway-ha/src/test/java/io/trino/gateway/ha/router/TestRoutingGroupSelector.java (hunk 5 above)
Comment (Chaho12) [reply gateway-ha/src/test/java/io/trino/gateway/ha/router/TestRoutingGroupSelector.java:433]: do you consider temporary tables (with clause) as non-tables?

File: gateway-ha/src/test/java/io/trino/gateway/ha/router/TestRoutingGroupSelector.java (hunk 3 above)
Comment (Chaho12) [reply gateway-ha/src/test/java/io/trino/gateway/ha/router/TestRoutingGroupSelector.java:414]: i don't think this is valid SQL. You cannot use uno as it is defined after dos.

```suggestion
                WITH dos AS (SELECT c1 from cat.schem.tbl1),
                uno as (SELECT c1 FROM dos)
                SELECT c1 FROM uno, dos
```

you can test this in trino via simple sql 

```sql
// works fine
WITH dos AS (SELECT node_id FROM system.runtime.nodes),
uno AS (SELECT node_id as node_id2 FROM dos)
SELECT node_id, node_id2
FROM uno, dos

// doesn't work as uno is not defined before dos.
// Query 20241023_021715_04215_f8xaz failed: line 1:34: Table 'system.runtime.uno' does not exist
WITH dos AS (SELECT node_id FROM uno),
uno AS (SELECT node_id as node_id2 FROM system.runtime.nodes)
SELECT node_id, node_id2
FROM uno, dos
```

**Comment by willmostly** [comment]
closing for #531
